streams/shared/src/main/scala/zio/stream/ZSummary.scala

```diff
@@ -0,0 +1,813 @@
+package zio.stream
+
+import scala.collection.mutable
+
+import zio._
+import zio.stream.internal.Builder
+
+/**
+ * A `ZSummary[R, E, I, O]` is a bounded aggregator of values of type `I`
+ * into values of type `O`, using an environment of type `R` and potential failures
+ * of type `E`.
+ *
+ * ZSummaries can be used to aggregate values on streams in several ways:
+ * - Summaries can be converted to transducers using [[ZSummary#toTransducer]] and
+ *   be used with [[ZStream#transduce]];
+ * - They can be converted to sinks using [[ZSummary#toSink]] and be
+ *   used with [[ZStream#run]];
+ * - They can be used to perform asynchronous, time-bounded aggregations with
+ *   [[[ZStream#aggregateAsync[R1<:R,E1>:E,P](summary*]]] and [[ZStream#aggregateAsyncWithin[R1<:R,E1>:E,P](summary*]];
+ * - They can be used directly with [[ZStream#scanWith]] and [[ZStream#zipWithSummary]].
+ *
+ * ZSummaries can be zipped together (using [[ZSummary#zipWith]]) to create composite
+ * aggregations and can be transformed using familiar combinators like [[ZSummary#map]],
+ * [[ZSummary#contramap]] and more.
+ *
+ * The majority of ZSummaries are pure: they do not use environmental effects and
+ * cannot fail. However, a select few can be effectful. It is recommended to stick
+ * to the pure ones for superior performance.
+ */
+sealed trait ZSummary[-R, +E, -I, +O] { self =>
```

Comment:
Is it possible/useful to allow arrow composition of summaries?

Reply:
Definitely useful. I gave it a try and ran into some trouble; I might have given up too fast but I thought it'd be fine to follow up with later.